### ISSUE TITLE
twisted.test.test_tcp.PortCleanerUpper is extremely confused

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10787]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#1181|
|Type|defect|
|Created|2005-09-19 16:35:44Z|
|Branch|https://github.com/twisted/twisted/tree/port-cleaner-upper-1181|

<details><summary>Searchable metadata</summary>

```
trac-id__1181 1181
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__branches_port_cleaner_upper_1181 branches/port-cleaner-upper-1181
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__tests tests
time__1127147744000000 1127147744000000
changetime__1203350925000000 1203350925000000
version__ 
owner__ 
cc__exarkun cc__spiv cc__itamarst cc__jml
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
PortCleanerUpper is broken:

  it calls loseConnection on ports it is given.  Ports do not have a
loseConnection method.  They have a stopListening method, and it can return a
Deferred.

  Clients are added to it.  Clients are not ports.  Since this seems like a
useful feature, it should have a separate list for clients (or connectors). 
Clients *do* have a loseConnection method.  However, it is difficult to know
when a client's connection has been lost.  This may make it difficult to get
timing right for this.

  cleanPorts tries to be synchronous.  It can't be synchronous, it should return
a Deferred.

  various tests throughout test_tcp are inconsistent in their use of cleanPorts.
 Some call it directly, some rely on tearDown to call it, some clean up their
own ports.  If PortCleanerUpper is a good idea, it should be used consistently
throughout the module.  Tests should add all their ports (and soon, clients) to
the lists.  tearDown should be allowed to call cleanPorts, and nothing else
should try to.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

tcp ports _do_ have a `stopListening` method. See [twisted/internet/tcp.py:802.](twisted/internet/tcp.py:802.)

cleanPorts is no longer synchronous (at least in deprecated-trial-1509) and probably in trunk.

The tests in test_tcp are dense, confusing and do use cleanPorts inconsistently.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

I think you misread.  I said tcp ports don't have a loseConnection method.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

Sorry. Brainfart.  But they _do_ have a `loseConnection` method, right? (line 788)?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

No.  It's an implementation detail.  Other implementations may not have ```loseConnection```.  The method the interface guarantees is ```stopListening```.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [22549]) Branching to 'port-cleaner-upper-1181'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

All the port cleaner upper objects are gone now.  I also made as many cleanups and improvements to the surrounding code as I could stomach.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set owner to @exarkun|
|-|-|

 * trailing whitespaces in test_tcp
 * at several places in test_tcp, the listen port number is assigned to an attribute (test_closePortInProtocolFactory, _connectedClientAndServerTest, test_connectByService, test_serverStartStop, test_clientStartStop). I don't think it's useful, it would be good to remove that.
 * there are a few assert_ in test_tcp that could be replaced with assertTrue
 * please open a ticket for removing loopUntil.

That's a huge cleanup, thanks for doing it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

Took care of those things.  The `loopUntil` ticket is [#3039](https://github.com/twisted/twisted/issues/3039).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set owner to @exarkun|
|-|-|

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

(In [22601]) Merge port-cleaner-upper-1181

Author: exarkun
Reviewer: therve
Fixes [#10787](https://github.com/twisted/twisted/issues/10787)

Remove several copies of a utility class for stopping listening ports in unit
tests and replace their usage with use of trial's new addCleanup method.
Also add docstring to a number of tests, convert some of them from using
loopUntil to using Deferreds and make other general cleanups.

Also fix the log messages and repr strings for certain cases involving new-style
classes.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
